pkg/controller/utils/elasticsearch.go

```diff
@@ -62,3 +103,271 @@ func GetElasticsearchClusterConfig(ctx context.Context, cli client.Client) (*ren
+	if es.client == nil {
+		user, password, root, err := getClientCredentials(client, ctx)
+		if err != nil {
+			return err
+		}
+		h := &http.Client{
+			Transport: &http.Transport{TLSClientConfig: &tls.Config{RootCAs: root}},
+		}
+
+		es.lock.Lock()
+		defer es.lock.Unlock()
+		esClient, err := newElastic(user, password, render.ElasticsearchHTTPSEndpoint, h)
+		if err != nil {
+			return err
+		}
+		es.client = esClient
+	}
+	return nil
+}
+
+// listILMPolicies generates ILM policies based on disk space and retention in LogStorage
+// Allocate 70% of ES disk space to flows, dns and bgp logs [majorPctOfTotalDisk]
+// Allocate 90% of the 70% ES disk space to flow logs, 5% of the 70% ES disk space to each dns and bgp logs.
+// Allocate 10% of ES disk space to logs that are NOT flows, dns or bgp [minorPctOfTotalDisk]
+// Equally distribute 10% of the ES disk space among these other log types
+func (es *EsClient) listILMPolicies(ls *operatorv1.LogStorage) map[string]policyDetail {
+	totalEsStorage := getTotalEsDisk(ls)
+	majorPctOfTotalDisk := 0.7
+
+	// numOfIndicesWithMinorSpace is the number of time series indices created that are not flows, dns or bgp related.
+	// i.e., audit_ee, audit_kube, compliance_reports, benchmark_results, events, snapshots
+	numOfIndicesWithMinorSpace := 6
+	minorPctOfTotalDisk := 0.1
+	pctOfDisk := minorPctOfTotalDisk / float64(numOfIndicesWithMinorSpace)
+
+	// Retention is not set in LogStorage for bgp, dns, benchmark and events logs, set default values used by curator
+	return map[string]policyDetail{
+		"tigera_secure_ee_flows": buildILMPolicy(totalEsStorage, majorPctOfTotalDisk, 0.9, int(*ls.Spec.Retention.Flows)),
+		"tigera_secure_ee_dns":   buildILMPolicy(totalEsStorage, majorPctOfTotalDisk, 0.05, 8),
+		"tigera_secure_ee_bgp":   buildILMPolicy(totalEsStorage, majorPctOfTotalDisk, 0.05, 8),
+
+		"tigera_secure_ee_audit_ee":           buildILMPolicy(totalEsStorage, minorPctOfTotalDisk, pctOfDisk, int(*ls.Spec.Retention.AuditReports)),
+		"tigera_secure_ee_audit_kube":         buildILMPolicy(totalEsStorage, minorPctOfTotalDisk, pctOfDisk, int(*ls.Spec.Retention.AuditReports)),
+		"tigera_secure_ee_snapshots":          buildILMPolicy(totalEsStorage, minorPctOfTotalDisk, pctOfDisk, int(*ls.Spec.Retention.Snapshots)),
+		"tigera_secure_ee_compliance_reports": buildILMPolicy(totalEsStorage, minorPctOfTotalDisk, pctOfDisk, int(*ls.Spec.Retention.ComplianceReports)),
+		"tigera_secure_ee_benchmark_results":  buildILMPolicy(totalEsStorage, minorPctOfTotalDisk, pctOfDisk, 91),
+		"tigera_secure_ee_events":             buildILMPolicy(totalEsStorage, minorPctOfTotalDisk, pctOfDisk, 91),
+	}
+}
+
+func (es *EsClient) createOrUpdatePolicies(ctx context.Context, listPolicy map[string]policyDetail) error {
+	for indexName, pd := range listPolicy {
+		policyName := indexName + "_policy"
+
+		res, err := es.client.XPackIlmGetLifecycle().Policy(policyName).Do(ctx)
```

Comment:
The docs of es library:
```
// XPackIlmPutLifecycle adds or modifies an ilm policy.
```
Does that mean we can just skip the Get Verb and go straight to the Put? That might save some code.

Reply:
XPackIlmPutLifecycle updates the policy and thereby updating the policy version
- I don't think we want the policy version number to be updated each time LogStorage reconciles, that might cause issues in log run where version hits whatever max value is allowed for that type.
- Also, ES notices that the policy version index is using is different than current policy version, so it has to do extra work https://www.elastic.co/guide/en/elasticsearch/reference/current/update-lifecycle-policy.html

So its better to split them up, we are not going to always call PUT, but yes it does involve making a GET call and comparing the values with LogStorage.

Reply:
Cool